crates/sparrow-materialize/src/materialize.rs

```diff
@@ -0,0 +1,235 @@
+            }
+
+            progress_tx
+                .send(MaterializationProgress::Stopped(last_progress))
+                .expect("TODO");
+            tracing::error!("materialization exited unexpectedly");
+            Ok(())
+        });
+
+        let status = MaterializationProgress::Uninitialized;
+        Self {
+            id: materialization.id.clone(),
+            stop_signal_sender: stop_tx,
+            progress_receiver: progress_rx,
+            handle: Some(handle),
+            status,
+        }
+    }
+
+    pub async fn stop(&mut self) -> error_stack::Result<(), Error> {
+        self.stop_signal_sender
+            .send(true)
+            .into_report()
+            .change_context(Error::Internal)?;
+
+        let handle = self.handle.take().expect("handle");
+        match &self.status {
+            MaterializationProgress::Uninitialized => {
+                // TODO: It's possible that the materialization process has started
+                // and progress has not been reported yet, so we'll report the incorrect
+                // status before stopping.
+                // We could fix this by reporting progress on shutdown, and having the
+                // handle return the latest progress.
+                match handle
+                    .await
+                    .into_report()
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status =
+                            MaterializationProgress::Stopped(ProgressInformation::default());
+                    }
+                    Err(e) => {
+                        self.status =
+                            MaterializationProgress::Failed(ProgressInformation::default(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Running(progress) => {
+                match handle
+                    .await
+                    .into_report()
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status = MaterializationProgress::Stopped(progress.clone());
+                    }
+                    Err(e) => {
+                        self.status = MaterializationProgress::Failed(progress.clone(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Stopped(_) => {
+                tracing::info!("cannot stop; materialization already stopped")
+            }
+            MaterializationProgress::Failed(_, _) => {
+                tracing::info!("cannot stop; materialization already failed")
+            }
+        };
+
+        Ok(())
+    }
+
+    pub fn get_status(&self) -> MaterializationStatus {
+        let progress = self.progress_receiver.borrow();
+        match &*progress {
```

Comment:
? weird pattern

Reply:
Yeah, is there a better way to do this? This was the only way I could deference the `MutexGuard` to get the value out, but I couldn't do it without the reference because it was moving values out of `self`, which is a reference.